broker/httpd/auth.go

```diff
@@ -23,61 +54,230 @@ func randomStringGeneration() (string, error) {
 	return base64.URLEncoding.EncodeToString(bytes), nil
 }
 
-func getRedirDestination(r *http.Request) string {
-	destinationPath := "/"
-	if !(r.Method == "GET" || r.Method == "POST") {
-		return destinationPath
-	}
-	err := r.ParseForm()
+func (s *Server) setAndStoreAuthCookie(w http.ResponseWriter, username string) error {
+	randomString, err := randomStringGeneration()
 	if err != nil {
-		return destinationPath
+		s.logger.Println(err)
+		return err
 	}
-	valueArr, ok := r.Form["returnURL"]
-	if !ok {
-		return destinationPath
+	expires := time.Now().Add(time.Hour * constants.CookieExpirationHours)
+	userCookie := http.Cookie{Name: authCookieName, Value: randomString, Path: "/", Expires: expires, HttpOnly: true, Secure: true}
+	http.SetCookie(w, &userCookie)
+	Cookieinfo := AuthCookie{username, userCookie.Expires}
+	s.cookieMutex.Lock()
+	s.authCookie[userCookie.Value] = Cookieinfo
+	s.cookieMutex.Unlock()
+	return nil
+}
+
+func (s *Server) getRedirURL(r *http.Request) string {
+	return "https://" + r.Host + constants.Oauth2redirectPath
+}
+
+func (s *Server) generateAuthCodeURL(state string, r *http.Request) string {
+	var buf bytes.Buffer
+	buf.WriteString(s.staticConfig.OpenID.AuthURL)
+	redirectURL := s.getRedirURL(r)
+	v := url.Values{
+		"response_type": {"code"},
+		"client_id":     {s.staticConfig.OpenID.ClientID},
+		"scope":         {s.staticConfig.OpenID.Scopes},
+		"redirect_uri":  {redirectURL},
 	}
 
-	inboundPath := valueArr[0]
-	if strings.HasPrefix(inboundPath, "/") {
-		destinationPath = inboundPath
+	if state != "" {
+		// TODO(light): Docs say never to omit state; don't allow empty.
+		v.Set("state", state)
 	}
+	if strings.Contains(s.staticConfig.OpenID.AuthURL, "?") {
+		buf.WriteByte('&')
+	} else {
+		buf.WriteByte('?')
+	}
+	buf.WriteString(v.Encode())
+	return buf.String()
+}
 
-	return destinationPath
+func (s *Server) generateValidStateString(r *http.Request) (string, error) {
+	key := []byte(s.staticConfig.Base.SharedSecrets[0])
+	sig, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.HS256, Key: key}, (&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
+		s.logger.Debugf(1, "New jose signer error err=%s", err)
```

Comment:
I'm still seeing multiple instances of this "err=%s" format. Please use this form instead:
`s.logger.Debugf(1, "New jose signer error: %s", err)`